### PR TITLE
Add instrumentation for Span decoration probes

### PR DESCRIPTION
**Pull request by jpbempel**
# What Does This Do
Introduced `CapturedContextInstrumentor` for sharing with Log & Span decoration probes.
Delegates commit phase of `CapturedContext` to probe based on the type for specific implementation.
Log probes: collect log message, eval condition, rate limit, collect captured contexts and collect eval errors
Span decoration probes: fetch current or root span and attach tags

# Motivation
Introduced new kind of probe to decorate existing span with dynamic tags

# Additional Notes
`DebuggerTransformerTest.testTransformations` is disabled and will be removed later
Parametric tests that is very long and cumbersome to maintain now with all refactorings done

### CODE REVIEW

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/LogInstrumentor.java [hunk 1]
```diff
@@ -61,793 +19,20 @@ public LogInstrumentor(
...
+    super(
+        logProbe,
+        classLoader,
+        classNode,
+        methodNode,
+        diagnostics,
+        probeIds,
+        logProbe.isCaptureSnapshot(),
+        toLimits(logProbe.getCapture()));
     this.capture = logProbe.getCapture();
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/LogInstrumentor.java:31]: we are using the first probe capture settings and not capture settings per probe.

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java [hunk 2]
```diff
@@ -0,0 +1,854 @@
...
+    getStatic(insnList, CAPTURED_CONTEXT_TYPE, "EMPTY_CONTEXT");
+    // stack [ret_value, capturedcontext]
+    insnList.add(new VarInsnNode(Opcodes.ASTORE, exitContextVar));
+    // stack [ret_value]
+    insnList.add(gotoNode);
+    return insnList;
+  }
+
+  @Override
+  protected InsnList getReturnHandlerInsnList() {
```
Comment (shatzi) [inline dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java:191]: do we still need this commit phase separated? from looking at Instrumentator class we simply push getBeforeReturnInsnList and then jump to the return handler, why not just write those two sections inside getBeforeReturnInsnList?

Comment (shatzi) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java:191]: it seems addFinallyHandler already doing that.

**Review by shatzi — APPROVED** [review]
Great progress!

See few minor issues and some refactor suggestions

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/SpanDecorationProbe.java [hunk 3]
```diff
@@ -83,14 +96,81 @@ public void instrument(
...
+    if (evaluateAt == MethodLocation.ENTRY || evaluateAt == MethodLocation.DEFAULT) {
+      status = entryContext.getStatus(id);
+    } else if (evaluateAt == MethodLocation.EXIT) {
+      status = exitContext.getStatus(id);
+    }
+    if (status == null) {
+      return;
+    }
+    AgentTracer.TracerAPI tracerAPI = AgentTracer.get();
+    AgentSpan agentSpan = tracerAPI.activeSpan();
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/SpanDecorationProbe.java:155]: yes true! 
done

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/SpanDecorationProbe.java [hunk 4]
```diff
@@ -83,14 +96,81 @@ public void instrument(
...
+    new SpanDecorationInstrumentor(this, classLoader, classNode, methodNode, diagnostics, probeIds)
+        .instrument();
+  }
+
+  @Override
+  public void evaluate(
+      Snapshot.CapturedContext context,
+      Snapshot.CapturedContext.Status status,
+      MethodLocation methodLocation) {
+    boolean shouldEvaluate = resolveEvaluateAt(methodLocation);
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/SpanDecorationProbe.java:109]: will try to move out, but need to expose in someway through ProbeDetails interface to access ProbeDefinition's evaluateAt. 
Will do in a next PR

File: dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java [hunk 5]
```diff
@@ -195,9 +197,7 @@ public enum Kind {
 
   /** Probe information associated with a snapshot */
   public interface ProbeDetails {
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java:199]: will do in a next PR

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/SpanDecorationProbe.java [hunk 6]
```diff
@@ -83,14 +96,81 @@ public void instrument(
...
+    if (status == null) {
+      return;
+    }
+    AgentTracer.TracerAPI tracerAPI = AgentTracer.get();
+    AgentSpan agentSpan = tracerAPI.activeSpan();
+    if (targetSpan == TargetSpan.ROOT) {
+      agentSpan = agentSpan.getLocalRootSpan();
+    }
+    List<Pair<String, String>> tagsToDecorate = status.getTagsToDecorate();
+    if (tagsToDecorate == null) {
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/SpanDecorationProbe.java:160]: true.
not typo this is the Interface name...

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/util/ValueScriptHelper.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+import static datadog.trace.bootstrap.debugger.util.TimeoutChecker.DEFAULT_TIME_OUT;
+
+import datadog.trace.bootstrap.debugger.Limits;
+import datadog.trace.bootstrap.debugger.Snapshot;
+import datadog.trace.bootstrap.debugger.util.TimeoutChecker;
+
+public class ValueScriptHelper {
+  private static final Limits LIMITS = new Limits(1, 3, 255, 5);
+
+  public static void serializeValue(
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/util/ValueScriptHelper.java:12]: not sure to understand, what do you want fo script evaluate (you mean conditions?)

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/LogInstrumentor.java (hunk 1 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/LogInstrumentor.java:31]: true. we need to merge/union all `Captures` of probes at same location.
Will do in a next PR

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java [hunk 8]
```diff
@@ -0,0 +1,854 @@
...
+  private final Limits limits;
+  private final LabelNode contextInitLabel = new LabelNode();
+  private int entryContextVar = -1;
+  private int exitContextVar = -1;
+  private int timestampStartVar = -1;
+  private int throwableListVar = -1;
+
+  public CapturedContextInstrumentor(
+      ProbeDefinition definition,
+      ClassLoader classLoader,
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java:58]: will do in a next PR

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/SpanDecorationProbe.java [hunk 9]
```diff
@@ -83,14 +96,81 @@ public void instrument(
...
+    for (Decoration decoration : decorations) {
+      try {
+        if (decoration.when != null) {
+          boolean condition = decoration.when.execute(context);
+          if (!condition) {
+            continue;
+          }
+        }
+        for (Tag tag : decoration.tags) {
+          Value<?> tagValue = tag.value.execute(context);
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/SpanDecorationProbe.java:122]: done

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/SpanDecorationProbe.java [hunk 10]
```diff
@@ -83,14 +96,81 @@ public void instrument(
...
+    }
+  }
+
+  @Override
+  public void commit(
+      Snapshot.CapturedContext entryContext,
+      Snapshot.CapturedContext exitContext,
+      List<Snapshot.CapturedThrowable> caughtExceptions) {
+    Snapshot.CapturedContext.Status status = null;
+    if (evaluateAt == MethodLocation.ENTRY || evaluateAt == MethodLocation.DEFAULT) {
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/probe/SpanDecorationProbe.java:146]: right
done

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java (hunk 2 above)
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java:191]: I think there are some constraint, but I can revisit this in a next PR

File: dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java [hunk 11]
```diff
@@ -0,0 +1,854 @@
...
+    }
+    insnList.add(new InsnNode(Opcodes.DUP));
+    // stack: [capturedcontext, capturedcontext]
+    ldc(insnList, fieldCount);
+    // stack: [capturedcontext, capturedcontext, int]
+    insnList.add(new TypeInsnNode(Opcodes.ANEWARRAY, CAPTURED_VALUE.getInternalName()));
+    // stack: [capturedcontext, capturedcontext, array]
+    int counter = 0;
+    List<FieldNode> fieldList = isStatic ? new ArrayList<>() : new ArrayList<>(classNode.fields);
+    if (correlationAvailable) {
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/src/main/java/com/datadog/debugger/instrumentation/CapturedContextInstrumentor.java:668]: yes can revisit this in a next PR

File: dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java [hunk 12]
```diff
@@ -765,6 +781,8 @@ public static class Status {
       boolean hasLogTemplateErrors;
       boolean hasConditionErrors;
       String message;
+      // Span decoration
+      List<Pair<String, String>> tagsToDecorate;
```
[the remark this answers is not in the document]
Comment (jpbempel) [reply dd-java-agent/agent-debugger/debugger-bootstrap/src/main/java/datadog/trace/bootstrap/debugger/Snapshot.java:785]: will do in a next PR

**Pull request merged**